File: 0881-boats-to-save-people/0881-boats-to-save-people.java
class Solution {
    public int numRescueBoats(int[] people, int limit) {
        //1. Initilialize the pointers 
        int start = 0;
        int end = people.length-1;
        int counter = 0;
        Arrays.sort(people);
        //2. while loop
        while(start<=end){
            //a. is the limit reached or less
            if(people[start]+people[end]<=limit){
                start++;
                end--;
                
            }
            //b. over exceeded the limit (only person in end can be accomodated)
            else{
                end--;
            }
            counter++;
            
        }
        return counter;
    }
}

File: 0881-boats-to-save-people/0881-boats-to-save-people.py
class Solution:
    # Greedy Solution
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        '''
        Lacks optimal substructure, can't build from child costs
        Time Complexity: O(nlgn)
        Space Complexity: O(1)
        '''
        
        # 1. Sort the array
        people.sort()
        
        #2. Initialize two pointers
        start = 0
        end = len(people)-1
        counter = 0
        #3. While loop 
        while(start<=end):
            #1. Check the sum if limit is reached
            if(people[start]+people[end]<=limit):
                start+=1
                end-=1
            else:
                end-=1
            counter+=1
        
        return counter
        